CarpetX/src/driver.hxx

```diff
@@ -152,13 +152,85 @@ struct GHExt {
   struct GlobalData {
     // all data that exists on all levels
 
+    class AnyTypeVector {
+    private:
+      int _type;
+      void *_data;
+      size_t _count;
+
+    public:
+      AnyTypeVector() : _type(-1), _data(nullptr), _count(0){};
+      AnyTypeVector(int type_, size_t count_)
+          : _type(type_),
+            _data(amrex::The_Arena()->alloc(CCTK_VarTypeSize(type_) * count_)),
+            _count(count_){};
+      // Noncopyable for now
+      AnyTypeVector(const AnyTypeVector &) = delete;
+      AnyTypeVector &operator=(const AnyTypeVector &) = delete;
+      AnyTypeVector &operator=(const AnyTypeVector &&other) {
+        this->_type = other._type;
+        this->_data = other._data;
+        this->_count = other._count;
+        return *this;
+      }
+      AnyTypeVector(AnyTypeVector &&other)
+          : _type(other._type), _data(other._data), _count(other._count) {}
+      void swap(AnyTypeVector &other) {
+        std::swap(this->_type, other._type);
+        std::swap(this->_data, other._data);
+        std::swap(this->_count, other._count);
+      }
+
+      ~AnyTypeVector() {
+        if (_data != nullptr) {
+          assert(_type != -1);
+          amrex::The_Arena()->free(_data);
+          _type = -1;
+          _data = nullptr;
+          _count = 0;
+        }
+        assert(_data == nullptr);
+        assert(_type == -1);
+        assert(_count == 0);
+      };
+
+      void alloc(int type_, size_t count_) {
+        assert(_type == -1);
+        assert(_data == nullptr);
+        assert(_count == 0);
+
+        _type = type_;
+        _data = amrex::The_Arena()->alloc(CCTK_VarTypeSize(type_) * count_);
+        _count = count_;
+      }
+
+      void free() {
+        assert(_type != -1);
+        assert(_data != nullptr);
+        amrex::The_Arena()->free(_data);
+        _type = -1;
+        _data = nullptr;
+        _count = 0;
+      }
+
+      int type() const { return _type; };
+
+      void *data_at(size_t i) const {
```

Comment:
There should probably be two versions of this function, one `const` and returning a `const void*`, and the other non-const and returning a `void *`.

Reply:
Done in 3605564. Note that differs from AMReX's `Array4D` and \`GridPtrDesc1\` class (and thus requires a const-cast somehwere in schedule.cxx)  
  
For grid-functions:  
\`\`\`  
    auto &restrict groupdata = \*leveldata.groupdata.at(gi);  
   const GridPtrDesc1 grid1(leveldata, groupdata, mfp);  
   for (int tl = 0; tl \< int(groupdata.mfab.size()); ++tl) {  
     const amrex::Array4\<CCTK\_REAL> vars =  
         groupdata.mfab.at(tl)->array(mfp.index());  
     for (int vi = 0; vi \< groupdata.numvars; ++vi)  
       cctkGH->data\[groupdata.firstvarindex + vi\]\[tl\] = grid1.ptr(vars, vi);  
   }  
\`\`\`  
  
For grid-arrays (and scalars):  
\`\`\`  
        auto &restrict arraygroupdata = \*globaldata.arraygroupdata.at(gi);  
       for (int tl = 0; tl \< int(arraygroupdata.data.size()); ++tl) {  
         const auto &restrict vars = arraygroupdata.data.at(tl);  
         for (int vi = 0; vi \< arraygroupdata.numvars; ++vi)  
           cctkGH->data\[arraygroupdata.firstvarindex + vi\]\[tl\] =  
               const\_cast\<void\*>(vars.data\_at(vi \* arraygroupdata.array\_size));  
\`\`\`